Location card: Various improvements and fixes

andrewfg opened the pull request:
This PR does the following..

1. The collection methods can now not only include exact Property / Equipment values but also (_**new**_) children of those values (optionally). This allows display of not only Lightbulb and Lightstripe, but also any LightSource or child thereof. And also of doors and windows that support either OpenState (digital) or OpenLevel (analog) positions; to name a couple.
2. Aligns the code in this PR with the [documentation](https://github.com/openhab/openhab-docs/pull/2514) and vice versa; taking into account [#556](https://github.com/openhab/openhab-webui/issues/556) and subsequent PRs related to that. This PR and #3239 added support for Motion detectors under presence, other displays under screens.
3. Cleans up the collection of lighting points to prevent duplicate entries in the location card; that may otherwise arise when (say) the user connects a Switch, a Dimmer, and a Color item to the same light channel.

Resolves https://github.com/openhab/openhab-webui/pull/3239#issuecomment-3004118138

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
Ping @mherwege

mherwege commented:
Everywhere we use `indexOf` to match with the tag and subtags, there is a risk.

Here is the code for equipment:
```
return arr.filter((e) => (partial) ? e.item.metadata.semantics.value.indexOf(value) === 0 : e.item.metadata.semantics.value === value)
```
When `partial` is true, this will match both `Equipment_Window` and `Equipment_WindowCovering` and their children, which is probably uninteded. The current use of the method avoid this, but some tag changes may still introduce this error.
The check should probably be on the parent tag and then on starting with parent tag + _.

andrewfg commented:
Good point. I will add a helper method for that.

andrewfg commented:
@mherwege I tried to update the documentation based on the code, but I then realised that I needed to update the code to match the documentation too. So can you please read this and cross reference with the current version of https://github.com/openhab/openhab-docs/pull/2514

mherwege commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -135,20 +135,22 @@ export default {
             ...findPoints(allPoints, 'Point_Status', false),
-            ...findPoints(allPoints, 'Point_Control', true)
+            ...findPoints(allPoints, 'Point_Control', true),
+            ...findPoints(allPoints, 'Point_Switch', false)
```
Switch is a child of Control I believe, so I don't think you need it explicitly if you check for the existence of Point_Control.

andrewfg replied:
As below..

mherwege commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -135,20 +135,22 @@ export default {
         case 'screens':
           equipment = [
-            ...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display', true),
+            ...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display_Television', true),
```
I am in doubt on this. What is the recommendation for the use of the Screen tag?

andrewfg replied:
I dunno. Someone in the review process insisted that a screen is something different than a display. And honestly I just rolled over and didn't fight it..

andrewfg replied:
So I include Television and Screen under screens. And Projector under projectors.

andrewfg replied:
Or we could double count Projector both under screens and under projectors?

andrewfg commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -163,7 +165,8 @@ export default {
             ...findPoints(allPoints, 'Point_Status', false, 'Property_Power'),
-            ...findPoints(allPoints, 'Point_Control', true, 'Property_Power')
+            ...findPoints(allPoints, 'Point_Control', false, 'Property_Power'),
+            ...findPoints(allPoints, 'Point_Switch', false, 'Property_Power')
```
Aha! I was overthinking it. I will revert.

andrewfg commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -135,20 +135,22 @@ export default {
             ...findPoints(allPoints, 'Point_Status', false, 'Property_Power'),
-            ...findPoints(allPoints, 'Point_Control', true, 'Property_Power')
+            ...findPoints(allPoints, 'Point_Control', false, 'Property_Power'),
+            ...findPoints(allPoints, 'Point_Switch', false, 'Property_Power')
```
As below

andrewfg commented:
@mherwege the one where I have the biggest doubt is lighting. Typically a light may have a Color channel which the binding would propose as Property_Color. And the user may link a Color control, a Dimmer control and a Switch to that channel. So by default all Items would have Property_Color and therefore excluded from the badge. So if the user wants the light to be included the badge they need to change just one of those Items to Property_Light (which one). Also all the Items will be (children of) Equipment_LightSource .. but probably that is not relevant. => Thoughts?

andrewfg commented:
@mherwege in light of https://github.com/openhab/openhab-webui/issues/556 and its subsequent issues I revised the documentation in https://github.com/openhab/openhab-docs/pull/2514 and the code here. And I think I recaptured the original intentions and documented them as such. However some notes:

1. I tweaked the code for lights -- the reason is based on my observations above in this thread, where a color lamp might have a switch item, an dimmer item and a color item linked to them, and we should not include all three such items in the collection.
 
2. I reverted the code for HVAC from Property.Any to Property.Power -- the reason is that many HVAC devices also have Property.Mode (e.g. heat/cool) or fan speed so we should only collect the Items that expressly relate to them being powered on.

3. I reverted the Point.Control to implicitly include children Point.Switch according to your comments above.

mherwege commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -222,11 +225,11 @@ export default {
       if (!equipment.length) return []
       let allPoints = allEquipmentPoints(equipment)
-      let points = findPoints(allPoints, 'Point', true, 'Property_Light')
+      let points = findPoints(allPoints, 'Point_Control_Switch', false)
```
In the previous code, as it does not look for a Switch, it will also find dimmer and color items. That is interpreted in line 203. Here it only looks for a switch.
I think we should look broader, but also avoid adding up multiple points from one equipment (e.g. having a Point_Control with a Dimmer item type and Point_Control_Switch with a Switch item type). In such case, the Switch item type should be used.

andrewfg replied:
> I think we should look broader, but also avoid adding up multiple points from one equipment (e.g. having a Point_Control with a Dimmer item type and Point_Control_Switch with a Switch item type). In such case, the Switch item type should be used.

I do basically agree with your statement :) .. however I don't know how to code it :( .. we would need to map each Item to the Thing that it comes from, and preferentially remove any duplications. => Perhaps you can kindly suggest some code?

mherwege replied:
> I do basically agree with your statement :) .. however I don't know how to code it :( .. we would need to map each Item to the Thing that it comes from, and preferentially remove any duplications. => Perhaps you can kindly suggest some code?

I need to look into it in more detail. I also don't think we should make it dependent on the thing (you could have 2 items essentially representing the same light, but linked to a different thing for some obscure reason). It should only be looking at the semantic relationships. What about prefering Point_Control_Switch over Point_Control? That may just do.

When you are done working on this, I will have a look to improve it.

mherwege commented:
In the following method, you changed from Point_Control to Point. It should probably be reverted to Point_Control as Point_Control_Switch would be covered as well.

```
    queryLightPoints () {
      let direct = findPoints(this.element.properties, 'Point', true, 'Property_Light')
      if (direct.length) return direct
      return findPoints(allEquipmentPoints(this.element.equipment), 'Point', true, 'Property_Light')
    }
```

mherwege commented:
> @mherwege the CI build reported 'Unexpected side effect in "overrideExpression" computed property' at line `#239` and Copilot suggested the fix which I pushed in the last commit. => However please check it.

That's my mistake. As the method had no arguments, I just moved it to a computed property with the idea of avoiding recalculation if nothing changes. But the side effect is needed.
So, it has to be moved back to a method like it was before, and all calls to it need the parantheses behind it to indicate a method is called (and not a computed property).

andrewfg commented:
> it has to be moved back to a method like it was before

Done.

mherwege commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -207,9 +208,29 @@ export default {
+        return []
+      }).flat())
+      // If there are no points, use the equipment items themselves
+      return equipment.filter((e) => e.points.length === 0).map((e) => e.item)
```
Looking at this again, my last commit (trying to remove the Set duplication removal), also removed the return statement for points. That should go back in.
Also equipment with no points (but an item) will only get added when there are no points at all now.
```suggestion
      points.push(...equipment.filter((e) => e.points.length === 0).map((e) => e.item))
      return points
```

mherwege commented:
@andrewfg Thanks for this. Someone else needs to further review now.

mherwege commented:
> Ok. Shall I revert my last two commits then?

That's my view on it indeed. You don't have to agree.
Before, it was not including them if there were LightSource equipments with (any) points. But I think they should be included in all cases if it is a LightSource equipment.

andrewfg commented:
@mherwege many thanks for guiding me through this one; this is the first time that I coded in JavaScript inside Vue, so I apologise for my lack of knowledge.

andrewfg commented:
@florian-h05 I noticed that you marked this as enhancement (which it certainly is) however does also have some aspect of fixing regression arising from recent changes in tag parent / child relationships. So could we please try to get it in v5.0 release?

https://community.openhab.org/t/openhab-5-0-milestone-discussion/162686/224

florian-h05 commented on bundles/org.openhab.ui/web/src/components/cards/glance/location/status-badge.vue:
```diff
@@ -126,22 +126,23 @@ export default {
           equipment = [
-            ...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display', true),
+            ...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display', false),
+            ...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display_Television', false),
```
Why is this explicitly needed?
Couldn't we use 

```js
...findEquipment(this.element.equipment, 'Equipment_AudioVisual_Display', true),
```

?

mherwege replied:
Projector is also a child of Display, and that has a separate badge. We didn’t want to break that.

florian-h05 replied:
Okay, fine then :+1:

florian-h05 reviewed (commented):
Code LGTM, one question though:
